File: 14/dec14.py
#!/usr/bin/env python

file = open("test.txt")
#file = open("input.txt")

template = ""
rules = []

nl = False
for line in file:
    if line == '\n':
        nl = True
        continue
    if not nl:
        template = line.strip()
    else:
        rules.append(line.strip().split(' -> '))

file.close()

def part1():
    polymer = template
    #print(polymer)
    steps = 40
    for s in range(steps):
        print(s, end='\r')
        p = ""
        for i in range(len(polymer)-1):
            pair = polymer[i] + polymer[i+1]
            #print(i, polymer, pair)
            for r in rules:
                if r[0] == pair:
                    p += pair[0] + r[1]
                    break
            #print(i, polymer, pair, p)
        polymer = p + pair[1]
        #print(polymer, len(polymer))
    print()
    chr = [0] * 26
    for c in polymer:
        i = ord(c) - ord('A')
        chr[i] += 1
    _min = 9e12
    _max = 0
    for c in chr:
        if c > 0:
            if c > _max:
                _max = c
            elif c < _min:
                _min = c
    print(_max - _min)
                




print("14 December 2021")
part1()
print()

File: 2021/14/dec14.go
package main

import (
	"bufio"
	"log"
	"os"
	"sort"
	"strings"
)

var rules map[string]string

func p1(tmpl string, rules map[string]string, steps int) {
	elements := make(map[string]int)
	pairs := make(map[string]int)
	for _, c := range tmpl {
		elements[string(c)]++
	}

	for i := 0; i < len(tmpl)-1; i++ {
		pairs[tmpl[i:i+2]]++
	}

	for i := 0; i < steps; i++ {
		newPairs := make(map[string]int)
		for pair, v := range pairs {
			e := rules[pair]
			elements[e] += v
			p1 := string(pair[0]) + string(e)
			p2 := string(e) + string(pair[1])
			newPairs[p1] += v
			newPairs[p2] += v
		}
		pairs = newPairs
	}

	var arr []int
	for _, e := range elements {
		arr = append(arr, e)
	}

	sort.Ints(arr)
	log.Println(arr[len(arr)-1] - arr[0])
}

func main() {
	log.Println("Day 14")
	//file, err := os.Open("test.txt")
	file, err := os.Open("input.txt")
	if err != nil {
		log.Fatal(err)
	}

	var template string
	rules = make(map[string]string)
	nl := false
	sc := bufio.NewScanner(file)
	for sc.Scan() {
		if len(sc.Text()) == 0 {
			nl = true
			continue
		}
		if nl {
			s := strings.Split(sc.Text(), " -> ")
			rules[s[0]] = s[1]
		} else {
			template = sc.Text()
		}
	}
	//log.Println(template)
	//log.Println(rules)
	p1(template, rules, 40)
}
